libraries/aws/provisioning/test/system/aws_iot_tests_provisioning_system.c

```diff
@@ -493,3 +648,73 @@ TEST( Provisioning_System, RegisterThingNominalCase )
     IotTest_Free( newCertificateContext.pCertificateIdBuffer );
     IotTest_Free( newCertificateContext.pCertificateOwnershipToken );
 }
+
+/**
+ * @brief Tests the behavior of the RegisterThing API function in the workflow when
+ * keys are generated by generated through the CreateCertificateFromCsr API
+ */
+TEST( Provisioning_System, RegisterThingWithCertFromCsrNominalCase )
+{
+    AwsIotProvisioningError_t status = AWS_IOT_PROVISIONING_SUCCESS;
+
+    AwsIotProvisioningRegisterThingCallbackInfo_t registerThingCallback =
+    {
+        .userParam = NULL,
+        .function  = _printRegisterThingResponseCallback
+    };
+
+    /* To test the Provisioning RegisterThing API, we need to request credentials from the Provisioning
+     * CreateCertificateFromCsr API, that */
+    /* we will use for provisioning in the test. */
+
+    _certIdAndOwnershipTokenContext_t newCertificateContext;
+
+    newCertificateContext.pCertificateIdBuffer = NULL;
+    newCertificateContext.certificateIdLength = 0;
+    newCertificateContext.pCertificateOwnershipToken = NULL;
+    newCertificateContext.tokenLength = 0;
+
+    AwsIotProvisioningCreateCertFromCsrCallbackInfo_t certFromCsrCallback =
+    {
+        .userParam = &newCertificateContext,
+        .function  = _storeDataOfCerFromCsrResponseCallback
+    };
+
+    /* Obtain new certificate and ownership token for testing provisioning API. */
+    status = AwsIotProvisioning_CreateCertificateFromCsr( _mqttConnection,
+                                                          IOT_MQTT_QOS_1,
+                                                          AWS_IOT_TEST_PROVISIONING_CSR_PEM,
+                                                          sizeof( AWS_IOT_TEST_PROVISIONING_CSR_PEM ),
+                                                          AWS_IOT_TEST_PROVISIONING_TIMEOUT,
+                                                          &certFromCsrCallback );
+
+    TEST_ASSERT_EQUAL( AWS_IOT_PROVISIONING_SUCCESS, status );
+
+
+    AwsIotProvisioningRegisterThingRequestInfo_t requestInfo;
+
+    requestInfo.pDeviceCertificateId = newCertificateContext.pCertificateIdBuffer;
```

Comment:
Is the malloc failed in _storeDataOfCerFromCsrResponseCallback() and this is NULL, and passed in AwsIotProvisioning_RegisterThing(), is that OK?

Reply:
Similar response as previous. (Added non-null checks)